src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp

```diff
@@ -0,0 +1,460 @@
+{
+    return GetAttribute(endpointId, Attributes::HardwareFaultAlert::Id, Attributes::HardwareFaultAlert::Get, hardwareFaultAlert);
+}
+
+bool SmokeCoAlarmServer::GetEndOfServiceAlert(EndpointId endpointId, EndOfServiceEnum & endOfServiceAlert)
+{
+    return GetAttribute(endpointId, Attributes::EndOfServiceAlert::Id, Attributes::EndOfServiceAlert::Get, endOfServiceAlert);
+}
+
+bool SmokeCoAlarmServer::GetInterconnectSmokeAlarm(EndpointId endpointId, AlarmStateEnum & interconnectSmokeAlarm)
+{
+    return GetAttribute(endpointId, Attributes::InterconnectSmokeAlarm::Id, Attributes::InterconnectSmokeAlarm::Get,
+                        interconnectSmokeAlarm);
+}
+
+bool SmokeCoAlarmServer::GetInterconnectCOAlarm(EndpointId endpointId, AlarmStateEnum & interconnectCOAlarm)
+{
+    return GetAttribute(endpointId, Attributes::InterconnectCOAlarm::Id, Attributes::InterconnectCOAlarm::Get, interconnectCOAlarm);
+}
+
+bool SmokeCoAlarmServer::GetContaminationState(EndpointId endpointId, ContaminationStateEnum & contaminationState)
+{
+    return GetAttribute(endpointId, Attributes::ContaminationState::Id, Attributes::ContaminationState::Get, contaminationState);
+}
+
+bool SmokeCoAlarmServer::GetSensitivityLevel(EndpointId endpointId, SensitivityEnum & sensitivityLevel)
+{
+    return GetAttribute(endpointId, Attributes::SensitivityLevel::Id, Attributes::SensitivityLevel::Get, sensitivityLevel);
+}
+
+chip::BitFlags<Feature> SmokeCoAlarmServer::GetFeatures(EndpointId endpointId)
+{
+    chip::BitFlags<Feature> featureMap;
+    if (!GetAttribute(endpointId, Attributes::FeatureMap::Id, Attributes::FeatureMap::Get, *featureMap.RawStorage()))
+    {
+        ChipLogError(Zcl, "Unable to get the Smoke CO Alarm feature map: attribute read error");
+        featureMap.ClearAll();
+    }
+    return featureMap;
+}
+
+/**********************************************************
+ * SmokeCoAlarmServer private methods
+ *********************************************************/
+
+bool SmokeCoAlarmServer::HandleRemoteSelfTestRequest(CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
+                                                     RemoteOpHandler opHandler)
+{
+    VerifyOrDie(nullptr != opHandler);
+
+    EndpointId endpointId = commandPath.mEndpointId;
+    Status status         = Status::Success;
+
+    ExpressedStateEnum expressedState;
+    bool success = GetExpressedState(endpointId, expressedState);
+
+    if (success)
+    {
+        // If the value is busy then return busy
+        if (expressedState == ExpressedStateEnum::kSmokeAlarm || expressedState == ExpressedStateEnum::kCOAlarm ||
+            expressedState == ExpressedStateEnum::kTesting || expressedState == ExpressedStateEnum::kInterconnectSmoke ||
+            expressedState == ExpressedStateEnum::kInterconnectCO)
+        {
+            status = Status::Busy;
+        }
+        else
+        {
+            success = opHandler(endpointId);
```

Comment:
Again: what is responsible for setting ExpressedState to "Testing"?  If it's the application then:

1. The example app being added here should actually do it!
2. That needs to be very clearly documented on the callback declaration.

Reply:
I have now changed to set ExpressedState at the server, so I have added the ExpressedState priority implementation.